src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,364 @@
+#include <exception>
+#include <gsl/gsl_errno.h>
+#include <gsl/gsl_matrix_double.h>
+#include <gsl/gsl_multiroots.h>
+#include <gsl/gsl_vector_double.h>
+#include <string>
+
+#include "ErrorHandling/Error.hpp"
+#include "Parallel/Printf.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace detail {
+template <size_t Dim, typename Solver>
+void print_state(size_t iteration_number, Solver* solver,
+                 bool print_header = false) noexcept {
+  if (print_header) {
+    Parallel::printf("Iter\t");
+    for (size_t i = 0; i < Dim; ++i) {
+      Parallel::printf(" x[%u]\t", i);
+    }
+    for (size_t i = 0; i < Dim; ++i) {
+      Parallel::printf(" f[%u]\t", i);
+    }
+    Parallel::printf("\n");
+  }
+
+  Parallel::printf("%u\t", iteration_number);
+  for (size_t i = 0; i < Dim; ++i) {
+    Parallel::printf("%3.4f  ", gsl_vector_get(solver->x, i));
+  }
+  for (size_t i = 0; i < Dim; ++i) {
+    Parallel::printf("%1.3e  ", gsl_vector_get(solver->f, i));
+  }
+  Parallel::printf("\n");
+}
+
+template <size_t Dim>
+std::array<double, Dim> gsl_to_std_array(const gsl_vector* const x) noexcept {
+  std::array<double, Dim> input_as_std_array{};
+  for (size_t i = 0; i < Dim; i++) {
+    gsl::at(input_as_std_array, i) = (gsl_vector_get(x, i));
+  }
+  return input_as_std_array;
+}
+
+template <size_t Dim>
+void gsl_vector_set_with_std_array(
+    gsl_vector* const func,
+    const std::array<double, Dim>& result_as_std_array) noexcept {
+  for (size_t i = 0; i < Dim; i++) {
+    gsl_vector_set(func, i, gsl::at(result_as_std_array, i));
+  }
+}
+
+template <size_t Dim>
+void gsl_matrix_set_with_std_array(
+    gsl_matrix* const matrix,
+    const std::array<std::array<double, Dim>, Dim>& matrix_array) noexcept {
+  for (size_t i = 0; i < Dim; i++) {
+    for (size_t j = 0; j < Dim; j++) {
+      gsl_matrix_set(matrix, i, j, gsl::at(gsl::at(matrix_array, i), j));
+    }
+  }
+}
+
+// The gsl_multiroot_function_fdf expects its functions to be of the form
+// int (* f) (const gsl_vector * x, void * params, gsl_vector * f).
+// However, we would like to be able to perform rootfinding on functions
+// of the form std::array<double, Dim> f(const std::array<double, Dim>& x).
+// So, we pass the function wrapper below to gsl_multiroot_function_fdf.
+// In the gsl documentation the third parameter is refered to as "void* params",
+// referring to the parameters that select out a particular function out of a
+// family of possible functions described in a class, but we instead pass the
+// pointer to the entire function object itself here. The
+// information of the function body is passed through with the Function template
+// parameter.
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_f(const gsl_vector* const x,
+                                       void* function_object,
+                                       gsl_vector* const f_of_x) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
```

Comment:
`fn` -> `typed_function_object` (or maybe it's better to rename `function_object` to `untyped_function_object`